cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,189 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+        info["repo"],
+        f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+    )
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(info["url"], info["ref"], download_path.path)
+
+    return download_path
+
+
+def _get_npm_dependencies(
+    download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, str]]
+) -> Dict[str, RootedPath]:
+    """
+    Download npm dependencies.
+
+    Receives the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param download_dir: Destination directory path where deps will be downloaded
+    :param deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dictionary of Resolved URL dependencies with downloaded paths
+    """
+    files_to_download: dict[str, dict[str, Any]] = {}
+    download_paths = {}
+    for url, info in deps_to_download.items():
+        if url.startswith("github:") or url.startswith("gitlab:") or url.startswith("bitbucket:"):
+            url = _update_vcs_url_with_full_hostname(url)
+
+        if url.startswith("git+"):
+            # Git source
+            download_paths[url] = _clone_repo_pack_archive(url, download_dir)
+        else:
+            if "registry.npmjs.org" == urlparse(url).hostname:
+                # Tar archive from npm registry
+                archive_name = f'{info["name"]}-{info["version"]}.tgz'.removeprefix("@").replace(
+                    "/", "-"
+                )
+                download_paths[url] = download_dir.join_within_root(archive_name)
+            else:
+                # Tar archive
+                if info["integrity"]:
+                    algorithm, digest = ChecksumInfo.from_sri(info["integrity"])
+                else:
+                    raise PackageRejected(
+                        f"{info['name']} is missing integrity checksum. It is mandatory"
+                        f"for https dependencies.",
+                        solution="Please double-check provided package-lock.json that"
+                        " your dependencies specify integrity. Try to "
+                        "rerun `npm install` on your repository.",
+                    )
+                download_paths[url] = download_dir.join_within_root(
+                    f"external-{info['name']}",
+                    f"{info['name']}-external-{algorithm}-{digest}.tgz",
+                )
+
+                # Create missing directories
+                directory = os.path.dirname(download_paths[url])
+                os.makedirs(directory, exist_ok=True)
+
+            files_to_download[url] = {
+                "download_path": download_paths[url],
+                "integrity": info["integrity"],
+            }
+
+    # Asynchronously download tar files
+    asyncio.run(
+        async_download_files(
+            {url: item["download_path"] for (url, item) in files_to_download.items()}, 5
```

Comment:
Is it worth adding a config param for the concurrency limit here?

Reply:
Yep, sounds good to me Bruno. I could update Config class https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/config.py#L11 ,  where I could add `concurrency_limit` and rename `gomod_download_max_tries` to `download_max_tries` and also stop using `DEFAULT_RETRY_OPTIONS` from http_request file -> https://github.com/containerbuildsystem/cachi2/pull/152/files/52386ead1bb932eb60b1e9bdafb611dfc1687351#diff-144107c22e96bfec985234d173328f1d9b1533859e60759e081bbe1a5fa0633cR117

How does it sound?

Reply:
Debatable on the `download_max_tries`. Gomod downloading is `go mod download`, retrying that is very different from http retries. I'm not against retrying everything the same amount of times, but then we would have to use that option for non-async http retries as well.

So maybe only `concurrency_limit` for now? :smile: